### PR TITLE
Fixed midiCC MSB message handling including UI updates

### PR DESCRIPTION
**Pull request by Vlietland**
Dear Jose, herewith my source code changes to enable midicc MSB handling. Since I am new in this repo, please check my changes thoroughly, to exclude possible introduced bugs. 

Many thanks and regards in advance for checking and hopefully accepting my changes. 

Jan

### CODE REVIEW

File: zyngui/zynthian_gui_layer.py [hunk 1]
```diff
@@ -507,8 +507,9 @@ def set_midi_chan_preset(self, midich, preset_index):
 					continue
 				if layer.set_preset(preset_index,True) and not selected:
 					try:
-						if not self.zyngui.modal_screen and self.zyngui.active_screen in ('control'):
-							self.select_action(self.root_layers.index(layer))
+						#following conditional statement removed to omit issue with not updating the screen
+						#if not self.zyngui.modal_screen and self.zyngui.active_screen in ('control'):
```
Comment (jofemodo) [inline zyngui/zynthian_gui_layer.py:511]: We don't want preset loading to change the active screen. What you see as a bug, it's a feature ;-)
Or perhaps i'm not understanding the problem you have.

File: zyngine/zynthian_layer.py [hunk 2]
```diff
@@ -1,13 +1,13 @@
 # -*- coding: utf-8 -*-
 #******************************************************************************
 # ZYNTHIAN PROJECT: Zynthian Layer (zynthian_layer)
-# 
+#
```
Comment (jofemodo) [inline zyngine/zynthian_layer.py:4]: Please, remove all "code formating" changes from the commit.
If you want to improve code format, then make a separated commit with the "code formatting" changes.

**Review by jofemodo — CHANGES_REQUESTED** [review]

**Comment by jofemodo** [comment]
Ahhh! Please, always PR to "testing" branch, never to "stable" branch.

Thanks

File: zyngine/zynthian_layer.py [hunk 3]
```diff
@@ -395,8 +395,16 @@ def midi_unlearn(self):
...
+		if ccnum==0:
+			logging.debug("Setting Bank: Receiving MIDI CH{}#CC{}={}".format(chan, ccnum, ccval))
+			self.set_bank(ccval)
+			self.load_preset_list()
+		elif ccnum==32:
+			#since preset is the same as program change, lsb needs to be ignored
+			#logging.debug("Setting Preset: Receiving MIDI CH{}#CC{}={}".format(chan, ccnum, ccval))
+			pass
+		elif self.engine:
+			#logging.debug("Receiving MIDI CH{}#CC{}={}".format(chan, ccnum, ccval))
```
Comment (jofemodo) [inline zyngine/zynthian_layer.py:407]: I propose this behaviour:
+ When CC0 (MSB) is received, we save it to a flag var "bank_msb". 
+ When CC32 (LSB) is received, we change the bank, using the "bank_msb" like this:

    + bank_msb==0 => system banks LSB (0-127)
    + bank_msb==1 => user bank LSB (0-127)
    + bank_msb==2 => favourites bank, if it exist. In this case, lsb has not meaning.

Of course, other schemes are possible, but i find this very simple and  easy to remember/understand.

File: zyngui/zynthian_gui_layer.py (hunk 1 above)
Comment (Vlietland) [reply zyngui/zynthian_gui_layer.py:511]: I proposed this change to omit the issue that the screen is not updated after processing the midi bank change message and a program control change message (instrument selection). For instance when the bank is updated the bank screen data shown is not changed. Same goes for the preset screen and likely also for the layer screen. This is very confusing for the user.

Do you have a low impact alternative solution to update the presently selected screen.

**Comment by jofemodo** [comment]
Closing this. I just implemented the feature in a separated branch using your initial idea & code:

https://github.com/zynthian/zynthian-ui/tree/midi_bank_change

It implements MSB like this:

+ 0 => system banks
+ 1 => user banks
+ 2 => external banks (USB drive)

LSB is the bank index. 

I improved the "refresh GUI" code, so now it should work fine. For instance, if active screen is "layers", it will get refreshed on bank/preset changes. If active screen is "bank" or "preset" list, it will jump to control. Etc.

Thanks a lot!

**Comment by Vlietland** [comment]
You are conveniently fast Jose. I had already implemented most of the MSB/LSB code myself and wanna to upload the solution to the test branch tonight. I will take a look at it and see if we implemented the same type of solution (for learning purposes).

**Pull request closed without merging**